File: packages/agent-python/shieldrasp/detection/sql.py
import sqlglot
from shieldrasp.taint.context import get_taint_context

def detect_sql_injection(query: str, dialect: str = "postgres"):
    if not ("union" in query.lower() or "1=1" in query):
        return {"blocked": False, "matched": False}
        
    try:
        expressions = sqlglot.parse(query, read=dialect)
        return {
            "blocked": True,
            "matched": True,
            "type": "SQL Injection",
            "confidence": 0.99,
            "cwe": "CWE-89",
            "payload": query
        }
    except Exception:
        return {
            "blocked": True,
            "matched": True,
            "type": "SQL Injection Error Based",
            "confidence": 0.85,
            "cwe": "CWE-89",
            "payload": query
        }

File: packages/agent-node/src/detection/sql.ts
import { TaintContext, getTaintContext } from '../taint/context';

const SQL_INJECTION_PATTERN = /UNION\s+SELECT|SLEEP\s*\(\d+\)|OR\s+['"]?1['"]?\s*=\s*['"]?1['"]?|DROP\s+TABLE|--|#|ALTER\s+TABLE/i;

export function detectSqlInjection(query: string, ctx?: TaintContext) {
    const matched = SQL_INJECTION_PATTERN.test(query);
    if (!matched) return { blocked: false, matched: false };

    const currentCtx = ctx || getTaintContext();
    if (!currentCtx) return { blocked: false, matched: false };

    // Simple heuristic: check if any part of the query contains a tainted value
    let tainted = false;
    for (const [taintedValue] of currentCtx.taintedObjects) {
        if (typeof taintedValue === 'string' && query.includes(taintedValue)) {
            tainted = true;
            break;
        }
    }

    if (tainted) {
        return {
            blocked: true, matched: true, attack_type: 'SQL Injection', confidence: 0.95,
            cwe: 'CWE-89', payload: query, timestamp: Date.now()
        };
    }
    return { blocked: false, matched: false };
}
